File: app/shared/src/agentSpec.ts
export const AGENTHUB_AGENT_SPEC_V1 = 'agenthub.agent_spec.v1' as const;

export interface AgentHubAgentSpecAvatarV1 {
  type: 'emoji' | 'icon';
  value: string;
}

export interface AgentHubAgentSpecRuntimeV1 {
  id: string;
  profile: string;
  provider: string;
  model: string;
  reasoning_effort?: string;
  temperature?: number;
  max_output_tokens?: number;
}

export interface AgentHubAgentSpecMCPServerV1 {
  id: string;
  transport: 'stdio' | 'http' | 'sse' | string;
  command?: string;
  url?: string;
}

export interface AgentHubAgentSpecMemoryPolicyV1 {
  mode: 'none' | 'project' | 'workspace' | string;
  retention: 'ephemeral-fixture' | 'session' | 'persistent' | string;
}

export interface AgentHubAgentSpecApprovalPolicyV1 {
  mode: string;
  require_approval_for?: string[];
}

export interface AgentHubAgentSpecTargetPreferenceV1 {
  mode: 'local-edge' | 'remote-edge' | 'hub-relay' | string;
  target_id?: string;
  health?: string;
}

export interface AgentHubAgentSpecFixturePolicyV1 {
  mode: 'fixture-only';
  no_spend: true;
  live_runtime_allowed: false;
}

export interface AgentHubAgentSpecV1 {
  schema_version: typeof AGENTHUB_AGENT_SPEC_V1;
  id: string;
  name: string;
  description: string;
  avatar: AgentHubAgentSpecAvatarV1;
  runtime: AgentHubAgentSpecRuntimeV1;
  skills: string[];
  mcp_servers: AgentHubAgentSpecMCPServerV1[];
  tool_allowlist: string[];
  memory_policy: AgentHubAgentSpecMemoryPolicyV1;
  approval_policy: AgentHubAgentSpecApprovalPolicyV1;
  target_preference: AgentHubAgentSpecTargetPreferenceV1;
  fixture: AgentHubAgentSpecFixturePolicyV1;
}

export interface AgentHubAgentSpecDraftV1 {
  id?: string;
  name: string;
  description?: string;
  avatar?: AgentHubAgentSpecAvatarV1;
  emoji?: string;
  runtimeId?: string;
  runtimeProfile: string;
  provider: string;
  model: string;
  reasoningEffort?: string;
  temperature?: number;
  maxOutputTokens?: number;
  skills?: string[];
  mcpServers?: AgentHubAgentSpecMCPServerV1[];
  toolAllowlist: string[];
  memoryPolicy?: AgentHubAgentSpecMemoryPolicyV1;
  approvalPolicy: {
    mode: string;
    requireApprovalFor?: string[];
  };
  targetPreference: {
    mode: string;
    targetId?: string;
    health?: string;
  };
}

export function buildAgentHubAgentSpecV1(draft: AgentHubAgentSpecDraftV1): AgentHubAgentSpecV1 {
  const name = draft.name.trim();
  const id = (draft.id ?? slugifyAgentSpecId(name)).trim();
  const runtimeId = firstNonEmpty(draft.runtimeId ?? '', normalizeRuntimeId(draft.runtimeProfile), 'custom-runtime');

  return {
    schema_version: AGENTHUB_AGENT_SPEC_V1,
    id,
    name,
    description: draft.description?.trim() ?? '',
    avatar: draft.avatar ?? { type: 'emoji', value: draft.emoji ?? 'A' },
    runtime: {
      id: runtimeId,
      profile: draft.runtimeProfile,
      provider: draft.provider,
      model: draft.model,
      ...(draft.reasoningEffort ? { reasoning_effort: draft.reasoningEffort } : {}),
      ...(typeof draft.temperature === 'number' ? { temperature: draft.temperature } : {}),
      ...(typeof draft.maxOutputTokens === 'number' ? { max_output_tokens: draft.maxOutputTokens } : {}),
    },
    skills: dedupeStrings(draft.skills ?? []),
    mcp_servers: draft.mcpServers ?? [],
    tool_allowlist: dedupeStrings(draft.toolAllowlist),
    memory_policy: draft.memoryPolicy ?? { mode: 'project', retention: 'ephemeral-fixture' },
    approval_policy: {
      mode: draft.approvalPolicy.mode,
      ...(draft.approvalPolicy.requireApprovalFor ? { require_approval_for: dedupeStrings(draft.approvalPolicy.requireApprovalFor) } : {}),
    },
    target_preference: {
      mode: draft.targetPreference.mode,
      ...(draft.targetPreference.targetId ? { target_id: draft.targetPreference.targetId } : {}),
      ...(draft.targetPreference.health ? { health: draft.targetPreference.health } : {}),
    },
    fixture: {
      mode: 'fixture-only',
      no_spend: true,
      live_runtime_allowed: false,
    },
  };
}

export function formatAgentHubAgentSpecV1(spec: AgentHubAgentSpecV1): string {
  return JSON.stringify(spec, null, 2);
}

function slugifyAgentSpecId(value: string): string {
  return value.trim().toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') || 'agenthub-agent-spec-fixture';
}

function normalizeRuntimeId(value: string): string {
  return value.trim().toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') || '';
}

function firstNonEmpty(...values: string[]): string {
  return values.find((value) => value.trim().length > 0) ?? '';
}

function dedupeStrings(values: string[]): string[] {
  return Array.from(new Set(values.map((value) => value.trim()).filter(Boolean)));
}

File: hub-server/internal/model/agent_spec.go
package model

import (
	"encoding/json"
	"fmt"
	"strings"
)

const agentHubAgentSpecV1Schema = "agenthub.agent_spec.v1"

type AgentHubAgentSpecV1 struct {
	SchemaVersion    string                   `json:"schema_version"`
	ID               string                   `json:"id"`
	Name             string                   `json:"name"`
	Description      string                   `json:"description,omitempty"`
	Avatar           map[string]string        `json:"avatar,omitempty"`
	Runtime          AgentSpecRuntimeV1       `json:"runtime"`
	Skills           []string                 `json:"skills,omitempty"`
	MCPServers       []AgentSpecMCPServerV1   `json:"mcp_servers,omitempty"`
	ToolAllowlist    []string                 `json:"tool_allowlist,omitempty"`
	MemoryPolicy     map[string]any           `json:"memory_policy,omitempty"`
	ApprovalPolicy   map[string]any           `json:"approval_policy,omitempty"`
	TargetPreference map[string]any           `json:"target_preference,omitempty"`
	Fixture          AgentSpecFixturePolicyV1 `json:"fixture,omitempty"`
}

type AgentSpecRuntimeV1 struct {
	ID              string  `json:"id"`
	Profile         string  `json:"profile"`
	Provider        string  `json:"provider"`
	Model           string  `json:"model"`
	ReasoningEffort string  `json:"reasoning_effort,omitempty"`
	Temperature     float64 `json:"temperature,omitempty"`
	MaxOutputTokens int64   `json:"max_output_tokens,omitempty"`
}

type AgentSpecMCPServerV1 struct {
	ID        string `json:"id"`
	Transport string `json:"transport"`
	Command   string `json:"command,omitempty"`
	URL       string `json:"url,omitempty"`
}

type AgentSpecFixturePolicyV1 struct {
	Mode               string `json:"mode"`
	NoSpend            bool   `json:"no_spend"`
	LiveRuntimeAllowed bool   `json:"live_runtime_allowed"`
}

// AgentProfileFromAgentSpecV1 projects a Builder export into Hub AgentProfile
// fixture input. The conversion keeps live runtime execution disabled and only
// prepares JSONB profile fields for tests/TeamRun fixture wiring.
func AgentProfileFromAgentSpecV1(ownerID string, spec AgentHubAgentSpecV1) (*AgentProfile, error) {
	if strings.TrimSpace(spec.SchemaVersion) != agentHubAgentSpecV1Schema {
		return nil, fmt.Errorf("unsupported AgentHubAgentSpec schema_version %q", spec.SchemaVersion)
	}
	if strings.TrimSpace(ownerID) == "" {
		return nil, fmt.Errorf("ownerID is required")
	}
	if strings.TrimSpace(spec.Name) == "" {
		return nil, fmt.Errorf("name is required")
	}
	if strings.TrimSpace(spec.Runtime.ID) == "" {
		return nil, fmt.Errorf("runtime.id is required")
	}
	if spec.Fixture.LiveRuntimeAllowed {
		return nil, fmt.Errorf("AgentProfile fixture conversion rejects live runtime allowance")
	}

	modelMapping := map[string]any{
		"schema_version":  spec.SchemaVersion,
		"runtime_profile": spec.Runtime.Profile,
		"provider":        spec.Runtime.Provider,
		"model":           spec.Runtime.Model,
		"memory_policy":   spec.MemoryPolicy,
		"avatar":          spec.Avatar,
		"fixture": map[string]any{
			"mode":                 spec.Fixture.Mode,
			"no_spend":             true,
			"live_runtime_allowed": false,
		},
	}

	modelMappingJSON, err := mustAgentSpecJSON(modelMapping)
	if err != nil {
		return nil, fmt.Errorf("marshal model_mapping: %w", err)
	}
	skillsJSON, err := mustAgentSpecJSON(spec.Skills)
	if err != nil {
		return nil, fmt.Errorf("marshal skills: %w", err)
	}
	mcpJSON, err := mustAgentSpecJSON(spec.MCPServers)
	if err != nil {
		return nil, fmt.Errorf("marshal mcp_servers: %w", err)
	}
	toolAllowlistJSON, err := mustAgentSpecJSON(spec.ToolAllowlist)
	if err != nil {
		return nil, fmt.Errorf("marshal tool_allowlist: %w", err)
	}
	approvalPolicyJSON, err := mustAgentSpecJSON(spec.ApprovalPolicy)
	if err != nil {
		return nil, fmt.Errorf("marshal approval_policy: %w", err)
	}
	targetPrefsJSON, err := mustAgentSpecJSON(spec.TargetPreference)
	if err != nil {
		return nil, fmt.Errorf("marshal target_preference: %w", err)
	}

	profile := &AgentProfile{
		OwnerID:                ownerID,
		Name:                   strings.TrimSpace(spec.Name),
		Description:            strings.TrimSpace(spec.Description),
		RuntimeID:              strings.TrimSpace(spec.Runtime.ID),
		Model:                  strings.TrimSpace(spec.Runtime.Model),
		Provider:               strings.TrimSpace(spec.Runtime.Provider),
		ReasoningEffort:        strings.TrimSpace(spec.Runtime.ReasoningEffort),
		ModelMapping:           modelMappingJSON,
		Skills:                 skillsJSON,
		MCPServers:             mcpJSON,
		ToolAllowlist:          toolAllowlistJSON,
		ApprovalPolicy:         approvalPolicyJSON,
		PermissionMode:         firstAgentSpecMapString(spec.ApprovalPolicy, "mode"),
		TargetPreferences:      targetPrefsJSON,
		ContextBudgetMaxTokens: int(spec.Runtime.MaxOutputTokens),
		Version:                1,
	}
	if profile.ReasoningEffort == "" {
		profile.ReasoningEffort = "medium"
	}
	if profile.PermissionMode == "" {
		profile.PermissionMode = "default"
	}
	if profile.ContextBudgetMaxTokens <= 0 {
		profile.ContextBudgetMaxTokens = 200000
	}
	if err := profile.Validate(); err != nil {
		return nil, err
	}
	return profile, nil
}

func mustAgentSpecJSON(value any) (string, error) {
	data, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func firstAgentSpecMapString(values map[string]any, key string) string {
	if values == nil {
		return ""
	}
	value, _ := values[key].(string)
	return value
}
